### PR TITLE
New feature : Remote control login by plugin

### PR DESCRIPTION
**Pull request by Shnoulle**
- Tested LDAP and DB
- Tested LDAP deactivated
- Tested LDAP with a DB user (OK : not loggued)
- Login form have a dropdown for selection, then allow to set it in get_session_key
- Maybe set plmugin to default by default ? But it can break current RC when update is OK
- webserver can not login via RC i think ?

Dis someone can test if it break webserver auth ? Really think not … but …

Removed old and uneeded  application/core/UserIdentity.php

### CODE REVIEW

File: application/libraries/PluginManager/AuthPluginBase.php [hunk 1]
```diff
@@ -67,6 +67,18 @@ public function afterLoginFormSubmit()
         }
     }
 
+    /**
+     * Set username and password by event
+     *
+     * @return null
+     */
+    public function remoteControlLogin()
```
Comment (Shnoulle) [inline application/libraries/PluginManager/AuthPluginBase.php]: New event … don't find another way . Maybe rename it ?

File: application/core/UserIdentity.php [hunk 2]
```diff
@@ -1,142 +0,0 @@
...
-* License: GNU/GPL License v2 or later, see LICENSE.php
-* LimeSurvey is free software. This version may have been modified pursuant
-* to the GNU General Public License, and as distributed it includes or
-* is derivative of works licensed under the GNU General Public License or
-* other free or open source software licenses.
-* See COPYRIGHT.php for copyright notices and details.
-*
-*/
-
-class UserIdentity extends CUserIdentity
```
Comment (Shnoulle) [inline application/core/UserIdentity.php]: Used (call) only by remote_control …

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 3]
```diff
@@ -2972,14 +2978,24 @@ public function get_uploaded_files($sSessionKey, $iSurveyID, $sToken)
      * @param string $sPassword password
      * @return bool
      */
-    protected function _doLogin($sUsername, $sPassword)
+    protected function _doLogin($sUsername, $sPassword,$sPlugin)
```
Comment (gekkedev) [inline application/helpers/remotecontrol/remotecontrol_handle.php]: Partly looks just like in https://github.com/LimeSurvey/LimeSurvey/pull/836, why didn't you add your commits to it?

**Comment by LouisGac** [comment]
great work, thank you very much Denis

**Comment by Shnoulle** [comment]
Let me check auto create user before … do you have a little hack for webserver Auth ? I can create one but if we have already one it can be great …

File: application/helpers/remotecontrol/remotecontrol_handle.php (hunk 3 above)
Comment (gekkedev) [reply application/helpers/remotecontrol/remotecontrol_handle.php]: If that's your preferred way, fine. However you as a contributor were probably able to edit the pull request to your needs whilst now I can not contribute but only comment lines or create pull requests for your fork, by creating a fork of your fork 👍

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 4]
```diff
@@ -2970,16 +2977,27 @@ public function get_uploaded_files($sSessionKey, $iSurveyID, $sToken)
...
+        $identity = new LSUserIdentity($sUsername, $sPassword);
+        $identity->setPlugin($sPlugin);
+        $event = new PluginEvent('remoteControlLogin');
+        $event->set('plugin', $sPlugin);
+        $event->set('username', $sUsername);
+        $event->set('password', $sPassword);
+        App()->getPluginManager()->dispatchEvent($event, array($sPlugin));
         if (!$identity->authenticate()) {
+            if( $identity->errorMessage ) {
+                // Be sure to don't return an empty string
```
Comment (gekkedev) [inline application/helpers/remotecontrol/remotecontrol_handle.php]: Better change this to "make sure to not return an empty string"

**Comment by Shnoulle** [comment]
Sh… auto create user still broken in 3.0.4 …

**Comment by Shnoulle** [comment]
Why `The condition is_string($loginResult) can never be true.` from [scrutinizer](https://scrutinizer-ci.com/g/LimeSurvey/LimeSurvey/inspections/83243696-dbda-4a72-bcad-0abc432b2677/issues/files/application/helpers/remotecontrol/remotecontrol_handle.php?status=new&orderField=path&order=asc&honorSelectedPaths=0) ?

Seems [OK](https://github.com/Shnoulle/LimeSurvey-Shnoulle/blob/b847f05c76e5570cba8c3211ee97c75273980d8d/application/helpers/remotecontrol/remotecontrol_handle.php#L2981) for me ?

**Comment by olleharstedt** [comment]
`_doLogin` returns boolean, so `is_string` is always false.

**Comment by Shnoulle** [comment]
Do login can return string here : https://github.com/LimeSurvey/LimeSurvey/pull/939/files#diff-1bcd4d2abe65594cf9ef157799417f85R2996 and it's php documented here https://github.com/LimeSurvey/LimeSurvey/pull/939/files#diff-1bcd4d2abe65594cf9ef157799417f85R2981

maybe it's better to add a Class variable $currentError …

**Comment by olleharstedt** [comment]
Yeah, I saw. That's why I deleted my comment. :D

**Comment by gekkedev** [comment]
@Shnoulle didn't you say autocreate was broken?

**Comment by Shnoulle** [comment]
@gekkedev fixed in https://github.com/LimeSurvey/LimeSurvey/commit/a5a05d59efb587120e0501792036bba51a7b4523

**Comment by c-schmitz** [comment]
@Shnoulle @olleharstedt From now on please PR new features only to develop branch, even small ones. Master does receive only bugfixes.

**Comment by olleharstedt** [comment]
Roger roger.

**Comment by gekkedev** [comment]
@c-schmitz https://github.com/LimeSurvey/LimeSurvey/pull/981 is a minor change to dev branch and still pending.

**Pull request merged**